lms/services/application_instance.py

```diff
@@ -116,29 +116,52 @@ def search(
 
         return query.all()
 
-    def build_from_lms_url(  # pylint:disable=too-many-arguments
-        self, lms_url, email, developer_key, developer_secret, settings
+    def create_application_instance(  # pylint:disable=too-many-arguments
+        self,
+        lms_url,
+        email,
+        developer_key,
+        developer_secret,
+        settings=None,
+        deployment_id=None,
+        lti_registration_id=None,
     ):
-        """Instantiate ApplicationInstance with lms_url."""
-        encrypted_secret = developer_secret
-        aes_iv = None
+        """Create an ApplicationInstance."""
+        consumer_key = (
+            "Hypothesis" + secrets.token_hex(16) if not deployment_id else None
+        )
 
         if developer_secret and developer_key:
             aes_iv = self._aes_service.build_iv()
             encrypted_secret = self._aes_service.encrypt(aes_iv, developer_secret)
+        else:
+            # If either one of developer_key or developer_secret is missing, then we
+            # don't save the other one either.
+            developer_key = encrypted_secret = developer_secret = aes_iv = None
+
+        try:
+            self._db.begin_nested()
+            application_instance = ApplicationInstance(
+                consumer_key=consumer_key,
+                shared_secret=secrets.token_hex(32),
+                lms_url=lms_url,
+                requesters_email=email,
+                developer_key=developer_key,
+                developer_secret=encrypted_secret,
+                aes_cipher_iv=aes_iv,
+                created=datetime.utcnow(),
+                settings=settings or {},
+                deployment_id=deployment_id,
+                lti_registration_id=lti_registration_id,
+            )
+            self._db.add(application_instance)
+            self._db.flush()  # Force the returned AI to have an ID
+        except IntegrityError:
+            # Rollback the failed transaction.
+            # Otherwise all following DB queries will fail
```

Comment:
Why is this something we want to carry on after? What's the situation that causes this that we are ok with?

Reply:
All access to the DB thru the same session won't work after the integrity error without the rollback.

The specific case here is accessing the DB after the error creating the instance to take you to the registration page, we do a LTIRegistration.get_by_id that fails otherwise.

I could get the registration here at the beginning of the method in all cases https://github.com/hypothesis/lms/blob/7291206ac4398fd7ef3216d00a94297b2a35734e/lms/views/admin/application_instance.py#L37

it will be a wasted query in the happy path but it's not like the admin pages are the bastion of performance. With that we can forget about all this business for now.

Reply:
I ended up doing this. No need for the rollback now.